### PR TITLE
[IMP] website, *: create new user and refactor tours

### PR DESCRIPTION
**Pull request by rahg-odoo**
*: test_website, website_sale, website_sale_stock_wishlist,
website_sale_stock

Specification:
- Use a less privileged user for tours that do not require admin
privileges.

Before this commit:
- Tours used admin privileges in most of the tours, even if features
do not require admin privileges.

After this commit:
- Created new internal user called "Rafe Restricted" with credentials
"username":"website_user", "password":"website_user"
- Created new class "HttpCaseWithWebsiteUser", this class can be
used anywhere we want to use restricted user login for tour.
- For existing tours, where admin privileges were not required login is
replaced with website_user.

task-3386375

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website/data/website_demo.xml [hunk 1]
```diff
@@ -1,5 +1,11 @@
 <?xml version="1.0" encoding="utf-8"?>
 <odoo>
+    <data>
```
Comment (msh-odoo) [inline addons/website/data/website_demo.xml:3]: Why we need separate data tag here?

File: odoo/addons/base/data/res_users_demo.xml [hunk 2]
```diff
@@ -16,6 +16,20 @@
...
+        <record id="partner_restricted" model="res.partner">
+            <field name="name">Rafe Restricted</field>
+            <field name="company_id" ref="main_company"/>
+            <field name="company_name">YourCompany</field>
+            <field name="street">725 5th Ave</field>
+            <field name="city">New York</field>
+            <field name="state_id"  model="res.country.state" search="[('code','=','NY')]"/>
+            <field name="zip">10022</field>
+            <field name="country_id" ref="us"/>
+            <field name="tz">Europe/Brussels</field>
```
Comment (msh-odoo) [inline odoo/addons/base/data/res_users_demo.xml:28]: Maybe keep NewYork timezone as country_id is USA

File: addons/website/data/website_demo.xml (hunk 1 above)
Comment (rahg-odoo) [reply addons/website/data/website_demo.xml:3]: There was no technical intentions, I thought rest of the data tags have similar kind of records, so instead adding into them I used another data tag. I saw same thing in accounts module. (/home/odoo/odoo/community/addons/account/demo/account_demo.xml)

**Comment by ryv-odoo** [comment]
What about the demo user then (Marc Demo) ? It matches your need of "Use a less privileged user for tours that do not require admin
privileges.", right ?

File: odoo/addons/base/data/res_users_demo.xml [hunk 3]
```diff
@@ -42,10 +56,24 @@
             <field name="image_1920" type="base64" file="base/static/img/user_demo-image.png"/>
         </record>
 
+        <record id="user_restricted" model="res.users">
```
Comment (kmagusiak) [inline odoo/addons/base/data/res_users_demo.xml:59]: Why not use `partner_demo` and give them the necessary permissions for the test?

**Comment by tde-banana-odoo** [comment]
Yup, please do not create demo data just for tests, we try to keep them business oriented and as simple as possible. Test users can be created in test setup anyway, you don't need demo nor data.

**Comment by kmagusiak** [comment]
> > What about the demo user then (Marc Demo) ? It matches your need of "Use a less privileged user for tours that do not require admin privileges.", right ?
> 
> It is mentioned in specifications that they need separate user just for that. only difference between marc demo and rafe restricted is that rafe restricted has "restricted editor" preset while marc demo has "editor and design" access.

During the setup of the test, you can add or remove permission from Marc Demo as you wish.

**Comment by HydrionBurst** [comment]
> it is mentioned in specifications that they need separate user just for that.
> only difference between marc demo and rafe restricted is that rafe restricted has "restricted editor" preset while marc demo has "editor and design" access.

I agree using an extra non-administrator user is better for your tour test, but it is tricky if someone wants to reuse the `restict` user in other teams. We call it `restrict` but it is hard to define how it is retricted. And logically all non adinistrator/super users are restrict users. I think the word `restrict` is not more informative than `Mark demo`. 🤔 
In the current code which we add groups in the demo data, but in the test we don't know if other groups are added to the user by other modules. Why not just create a `HttpCaseWithWebsiteUser` for your team?

**Comment by msh-odoo** [comment]
> HttpCaseWithWebsiteUser

I do agree on ```HttpCaseWithWebsiteUser``` Instead of creating user in demo we can setup user here with restricted editor and use this ```HttpCaseWithWebsiteUser``` wherever needed.

**Comment by rahg-odoo** [comment]
> > HttpCaseWithWebsiteUser
> 
> I do agree on `HttpCaseWithWebsiteUser` Instead of creating user in demo we can setup user here with restricted editor and use this `HttpCaseWithWebsiteUser` wherever needed.

Okay!
- I will remove demo data and create user only in test environment. 
- Change name of user as restricted is broad term.

**Review by SergeBayet — CHANGES_REQUESTED** [review]
LGTM 👍 Thanks! I have a comment below regarding the creation of the restricted user.

Are we certain that the test-tour runs as admin when it targets an admin-only feature? Otherwise, everything else should use the restricted editor. How did you compile that list? Just curious! 🙂

**Comment by rahg-odoo** [comment]
> LGTM 👍 Thanks! I have a comment below regarding the creation of the restricted user.
> 
> Are we certain that the test-tour runs as admin when it targets an admin-only feature? Otherwise, everything else should use the restricted editor. How did you compile that list? Just curious! 🙂

@SergeBayet No, everything else should not use restricted editor. whichever tour does not require `editor and snippet options` during tour and does not dragging or dropping snippet should use this newly created user.

This is how things look like for that user when he opens up editor.
image

I created that list with manual efforts...checking each tour and running with login "website_user" if needed.
Thanks for the review!

File: addons/website/tests/test_client_action.py [hunk 4]
```diff
@@ -20,7 +21,7 @@ def test_01_client_action_redirect(self):
             'url': '/test_client_action_redirect',
             'is_published': True,
         })
-        self.start_tour(page.url, 'client_action_redirect', login='admin')
+        self.start_tour(page.url, 'client_action_redirect', login="website_user")
 
     def test_02_client_action_iframe_fallback(self):
         self.start_tour('/@/', 'client_action_iframe_fallback', login='admin')
```
[the remark this answers is not in the document]
Comment (rahg-odoo) [reply addons/website/tests/test_client_action.py:27]: @SergeBayet To decide that, assume it does not require admin rights and try running tour with "website_user" if tour fails assumption was wrong and it does require admin rights. In this case tour failed, so it requires admin rights.

**Comment by SergeBayet** [comment]
You can remove 'base' from the module list in your commit message and the PR description since it hasn't been modified.

File: addons/website/tests/common.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+            'image_1920': base64.b64encode(file_open("website/static/src/img/user-restricted-image.png", "rb").read()),
+        }
+        cls.user_website_user = cls.env['res.users'].create(user_vals)
+        cls.user_website_user.write({
+            'groups_id': [
+                (3, cls.env.ref('website.group_website_designer').id),
+                (4, cls.env.ref('website.group_website_restricted_editor').id),
+                (4, cls.env.ref('base.group_user').id)
+            ]
+        })
```
[the remark this answers is not in the document]
Comment (rahg-odoo) [reply addons/website/tests/common.py]: @SergeBayet
Done. I made a slight adjustment to your suggested code, as we are not setting all three groups. Instead, we are removing one and adding two groups.

[the remark this answers is not in the document]
Comment (SergeBayet) [reply addons/website/tests/common.py]: Yep, my bad. Thanks

**Comment by SergeBayet** [comment]
@rahg-odoo I don't know why the `ci/upgrade_entreprise` failed. Maybe a rebase on master will be enough.

**Comment by SergeBayet** [comment]
@rahg-odoo Don't forget the title of the PR description that still contains "base" module.
image

**Comment by bso-odoo** [comment]
@robodoo delegate=SergeBayet

**Comment by SergeBayet** [comment]
@robodoo r+

**Pull request closed without merging**